src/lsm/manifest_log.zig

```diff
@@ -954,47 +1000,237 @@ pub fn ManifestLogType(comptime Storage: type) type {
+
+        // "limit of MC(c) as c approaches ∞":
+        // Working out this recurrence relation's limit with a closed-form solution is complicated.
+        // Just compute the limit iteratively instead. (1024 is an arbitrary safety counter.)
+        var log_blocks_before: u32 = 0;
+        const log_blocks_cycle_max = for (0..1024) |_| {
+            const log_blocks_after =
+                log_blocks_full_max +
+                half_bar_append_blocks_max *
+                stdx.div_ceil(log_blocks_before, half_bar_compact_blocks_max);
+
+            if (log_blocks_before == log_blocks_after) {
+                break log_blocks_after;
+            }
+            log_blocks_before = log_blocks_after;
+        } else {
+            // If the value does not converge within the given number of steps,
+            // constants.lsm_manifest_compact_blocks_extra should probably be raised.
+            @panic("ManifestLog.Pace.log_blocks_cycle_max: no convergence");
+        };
+
+        const log_blocks_burst_max = half_bar_append_blocks_max *
+            stdx.div_ceil(log_blocks_full_max + 1, half_bar_compact_blocks_max);
+
+        // "limit of MB(b) as b approaches ∞":
+        const log_blocks_max = log_blocks_cycle_max + log_blocks_burst_max;
 
-    fn blocks_count_appends(options: *const Options) usize {
-        return stdx.div_ceil(options.compaction_appends_max(), schema.ManifestNode.entry_count_max);
+        return .{
+            .half_bar_append_blocks_max = half_bar_append_blocks_max,
+            .half_bar_compact_blocks_max = half_bar_compact_blocks_max,
+            .log_blocks_full_max = log_blocks_full_max,
+            .log_blocks_max = log_blocks_max,
+            .log_blocks_cycle_max = log_blocks_cycle_max,
+        };
     }
 
-    /// The upper-bound of manifest blocks we must buffer.
-    ///
-    /// `blocks` must have sufficient capacity for:
-    /// - a manifest log compaction (+1 block in the worst case)
-    /// - a leftover open block from the previous ops (+1 block)
-    /// - table updates from a half bar of compactions
-    fn blocks_count_max(options: *const Options) usize {
-        const count = 1 + 1 + options.blocks_count_appends();
-        assert(count >= 3);
-
-        return count;
+    fn half_bar_compact_blocks(pace: Pace, options: struct {
+        /// The number of manifest blocks that *currently* exist.
+        log_blocks_count: u32,
+    }) u32 {
+        if (options.log_blocks_count < pace.log_blocks_cycle_max) {
+            // We have enough free manifest blocks that we could go a whole "cycle" without
+            // compacting any. It doesn't strictly matter how much compaction we do in this case, so
+            // just try to pace the work evenly, erring on the side of doing less work than
+            // necessary.
+            return @divFloor(
+                pace.half_bar_compact_blocks_max * options.log_blocks_count,
+                pace.log_blocks_cycle_max,
+            );
```

Comment:
This is super hand-wavy as I don't yet fully understand the logic here, but dont' we need an extra `option` here? It feels like, to pace compaction, we need to know not only the size of the manifest (`log_blocks_count`), but also the number of tables?

Consider this two situations:

|`log_blocks_count` | `table_count` |
-|-
|  100              |   10          |
|  100              |   1000        |

It feels like we should compact quite a bit in the first case (few tables), but maybe not as much in the second case. 

The current logic can't do that, because it has access only to the first column of this table. 

In other words, the calculation we do are relative to `T` --- the maximum number of tables that could in theory be supported by the data file. But in practice for the vast majority of data files would have some actual number of tables `t`, such that `t ≪ T`. 

It's not clear to me what would be the behavior of our pacing in this, non-extreme case? I think it could be one of:

* the manifest log grows to its maximal extent (as if it has full T tables)
* the manifest log shrinks to something very small
* the manifest log maintains maintains the same load factor as it would have in t=T case

I _think_ the behavior we want here is the last one --- if your data file is small, we want the log to be small as well (speedier recovery), but also we don't want to waste compaction work churning through "live" tables. I would be a bit surprised if we can achieve that behavior without passing the current number of tables, but maybe that's how it works out?

Reply:
Good point!

So in short we to pace compaction relative to the log's density (the ratio of live tables to total entries).
Though at the same time, when the absolute number of log blocks is low, even if it is low density we probably don't want to compact at the maximum rate.

This feels more complicated than it "should" be, but:

```zig
            // We have enough free manifest blocks that we could go a whole "cycle" without
            // compacting any. It doesn't strictly matter how much compaction we do in this case, so
            // just try to pace the work evenly, maintaining a constant load factor with respect to
            // the cycle-max.

            // Our "target" block count extrapolates a log block count from our table count and the
            // log's maximum load factor.
            const log_blocks_target = @max(1, @divFloor(
                pace.log_blocks_cycle_max * options.tables_count,
                pace.tables_max,
            ));

            // Scale the compaction rate according to the total number of blocks.
            // e.g. if our manifest log only has 10 blocks, we shouldn't compact at the same
            // rate we would if we had 1000 blocks, even if the tables are low density.
            const compact_blocks_limit = @min(
                pace.log_blocks_cycle_max,
                stdx.div_ceil(
                    pace.half_bar_compact_blocks_max * options.log_blocks_count,
                    pace.log_blocks_cycle_max,
                ),
            );

            // Finally, scale the compaction rate based on the current load factor.
            return @min(
                compact_blocks_limit,
                @divFloor(
                    compact_blocks_limit * options.log_blocks_count,
                    log_blocks_target,
                ),
            );
```

Reply:
SGTM! 

I think that maybe scaling by the total number of blocks is not _that_ necessary though --- if there are few blocs, but the load factor is high, then compaction is probably going to reduce the load factor significantly, as, for short manifest, we compact larger fraction of the manifest. 

If we don't scale by manifest log size, we can look _just_ at the load factor, and that'll naturally handle the case where we exceed pace.log_blocks_cycle_max, which would be simplification.

Reply:
> I think that maybe scaling by the total number of blocks is not that necessary though


In that case I'd rather simplify!:

```zig
        // We have enough free manifest blocks that we could go a whole "cycle" without
        // compacting any. It doesn't strictly matter how much compaction we do in this case, so
        // just try to pace the work evenly, maintaining a constant load factor with respect to
        // the cycle-max.

        // Our "target" block count extrapolates a log block count from our table count and the
        // log's maximum load factor.
        const log_blocks_target = @max(1, @divFloor(
            pace.log_blocks_cycle_max * options.tables_count,
            pace.tables_max,
        ));

        return @min(
            pace.half_bar_compact_blocks_max,
            @divFloor(
                pace.half_bar_compact_blocks_max * options.log_blocks_count,
                log_blocks_target,
            ),
        );
```

i.e. scale `half_bar_compact_blocks_max` to the ratio of "current log blocks" to "target log blocks".

> If we don't scale by manifest log size, we can look just at the load factor, and that'll naturally handle the case where we exceed pace.log_blocks_cycle_max, which would be simplification.

I think I like handling the

```zig
        // Pretend we have an extra half_bar_append_blocks_max blocks so that we always switch to
        // the maximum compaction rate before we exceed the cycle-max.
        if (pace.log_blocks_cycle_max <=
            options.log_blocks_count + pace.half_bar_append_blocks_max)
        {
            return pace.half_bar_compact_blocks_max;
        }
```
branch explicitly + independently though, since it is critical for the upper-bound calculation, whereas the other branch is "anything goes".

---

So all together:

```zig
    fn half_bar_compact_blocks(pace: Pace, options: struct {
        /// The number of manifest blocks that *currently* exist.
        log_blocks_count: u32,
        /// The number of live tables.
        tables_count: u32,
    }) u32 {
        assert(options.tables_count <= pace.tables_max);

        // Pretend we have an extra half_bar_append_blocks_max blocks so that we always switch to
        // the maximum compaction rate before we exceed the cycle-max.
        if (pace.log_blocks_cycle_max <=
            options.log_blocks_count + pace.half_bar_append_blocks_max)
        {
            return pace.half_bar_compact_blocks_max;
        }

        // We have enough free manifest blocks that we could go a whole "cycle" without
        // compacting any. It doesn't strictly matter how much compaction we do in this case, so
        // just try to pace the work evenly, maintaining a constant load factor with respect to
        // the cycle-max.

        // Our "target" block count extrapolates a log block count from our table count and the
        // log's maximum load factor.
        const log_blocks_target = @max(1, @divFloor(
            pace.log_blocks_cycle_max * options.tables_count,
            pace.tables_max,
        ));

        return @min(
            pace.half_bar_compact_blocks_max,
            @divFloor(
                pace.half_bar_compact_blocks_max * options.log_blocks_count,
                log_blocks_target,
            ),
        );
    }
```

Reply:
> I think I like handling the branch explicitly

:+1: I was also thinking that at the back of my mind!